File: app/src/main/java/com/niiadotei/storehouse/ui/apprentice/ApprenticeFragment.java
package com.niiadotei.storehouse.ui.apprentice;

import android.app.DatePickerDialog;
import android.content.res.Resources;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatButton;
import androidx.fragment.app.Fragment;

import com.niiadotei.storehouse.data.DatabaseHelper;
import com.niiadotei.storehouse.databinding.FragmentApprenticeBinding;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

public class ApprenticeFragment extends Fragment{
    private FragmentApprenticeBinding binding;

    DatabaseHelper databaseHelper;

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(true);
    }

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentApprenticeBinding.inflate(inflater, container, false);

        Resources resources = this.getResources();
        Locale locale = resources.getConfiguration().locale;
        NumberFormat currencyInstance = DecimalFormat.getCurrencyInstance(locale);

        String date = new SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(new Date());

        TextView pastSalesMadeTextView = binding.pastSalesMadeTextView;
        TextView pastSalesQuantityTextView = binding.pastSalesQuantityTextView;
        TextView pastSalesNumberTextView = binding.pastSalesNumberTextView;
        TextView pastCustomerNumberTextView = binding.pastCustomerNumberTextView;
        TextView dateOfPastSalesTextView = binding.dateOfPastSales;

        TextView salesPerformanceValueTextView = binding.salesPerformanceValue;
        TextView salesPerformancePercentTextView = binding.salesPerformancePercent;
        TextView differenceInQuantityTextView = binding.differenceInQuantitySold;

        databaseHelper = new DatabaseHelper(this.getActivity());

        ApprenticeRetriever apprenticeRetriever = new ApprenticeRetriever(date, this.requireContext());

        AppCompatButton salesDateButton = binding.salesDateButton;

        TextView totalSalesMadeTextView = binding.salesMadeTextView;
        double totalSalesMade = apprenticeRetriever.getTotalSalesMadeToday();
        totalSalesMadeTextView.setText(currencyInstance.format(totalSalesMade));

        TextView salesQuantityTextView = binding.salesQuantityTextView;
        salesQuantityTextView.setText(String.valueOf(apprenticeRetriever.getTotalQuantitySoldToday()));

        TextView salesNumberTextView = binding.salesNumberTextView;
        salesNumberTextView.setText(String.valueOf(apprenticeRetriever.getTotalNumberOfSalesToday()));

        TextView customerNumberTextView = binding.customerNumberTextView;
        customerNumberTextView.setText(String.valueOf(apprenticeRetriever.getNumberOfNewCustomersToday()));

        DatePickerDialog.OnDateSetListener dateSetListener = (datePicker, year, month, day) -> {
            Calendar calendar = Calendar.getInstance();

            calendar.set(Calendar.YEAR, year);
            calendar.set(Calendar.MONTH, month);
            calendar.set(Calendar.DAY_OF_MONTH, day);

            String pastDate = new SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(calendar.getTime());

            ApprenticeRetriever apprenticeRetrieverPast = new ApprenticeRetriever(pastDate, requireContext());

            String dateOfPastSales = "Date of Sales: " + pastDate;
            dateOfPastSalesTextView.setText(dateOfPastSales);

            double pastTotalSalesMade = apprenticeRetrieverPast.getTotalSalesMadeToday();
            pastSalesMadeTextView.setText(currencyInstance.format(pastTotalSalesMade));

            pastSalesQuantityTextView.setText(String.valueOf(apprenticeRetrieverPast.getTotalQuantitySoldToday()));

            pastSalesNumberTextView.setText(String.valueOf(apprenticeRetrieverPast.getTotalNumberOfSalesToday()));

            pastCustomerNumberTextView.setText(String.valueOf(apprenticeRetrieverPast.getNumberOfNewCustomersToday()));

            double salesPerformanceValue = totalSalesMade - pastTotalSalesMade;
            salesPerformanceValueTextView.setText(currencyInstance.format(salesPerformanceValue));

            DecimalFormat decimalFormat = new DecimalFormat("#.##");
            double salesPerformancePercent = 0.0;

            if (totalSalesMade != 0.0) {
                salesPerformancePercent = Double.parseDouble(decimalFormat.format(((totalSalesMade - pastTotalSalesMade)  / totalSalesMade) * 100));
            }

            if (Double.isNaN(salesPerformancePercent)) {
                salesPerformancePercent = 0.0;
            }

            String salesPerformancePercentString = String.format(Locale.US, "%.0f%%", salesPerformancePercent);
            salesPerformancePercentTextView.setText(salesPerformancePercentString);

            int differenceInQuantity = apprenticeRetriever.getTotalQuantitySoldToday() - apprenticeRetrieverPast.getTotalQuantitySoldToday();
            differenceInQuantityTextView.setText(String.valueOf(differenceInQuantity));
        };

        salesDateButton.setOnClickListener(view -> {
            Calendar calendar = Calendar.getInstance();

            int year = calendar.get(Calendar.YEAR);
            int month = calendar.get(Calendar.MONTH);
            int day = calendar.get(Calendar.DAY_OF_MONTH);

            DatePickerDialog datePickerDialog = new DatePickerDialog(getActivity(), dateSetListener, year, month, day);
            datePickerDialog.show();

        });

        return binding.getRoot();
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/niiadotei/storehouse/ui/apprentice/ApprenticeFragment.kt
package com.niiadotei.storehouse.ui.apprentice

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener

import android.os.Bundle

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import android.widget.DatePicker
import android.widget.TextView

import androidx.fragment.app.Fragment

import com.niiadotei.storehouse.data.DatabaseHelper
import com.niiadotei.storehouse.databinding.FragmentApprenticeBinding

import java.text.DecimalFormat
import java.text.SimpleDateFormat

import java.util.*

class ApprenticeFragment : Fragment() {
    private lateinit var binding: FragmentApprenticeBinding

    var databaseHelper: DatabaseHelper? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentApprenticeBinding.inflate(inflater, container, false)

        val resources = this.resources
        val locale = resources.configuration.locale
        val currencyInstance = DecimalFormat.getCurrencyInstance(locale)

        val date = SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(Date())

        val pastSalesMadeTextView: TextView = binding.pastSalesMadeTextView
        val pastSalesQuantityTextView: TextView = binding.pastSalesQuantityTextView
        val pastSalesNumberTextView: TextView = binding.pastSalesNumberTextView
        val pastCustomerNumberTextView: TextView = binding.pastCustomerNumberTextView
        val dateOfPastSalesTextView: TextView = binding.dateOfPastSales
        val salesPerformanceValueTextView: TextView = binding.salesPerformanceValue
        val salesPerformancePercentTextView: TextView = binding.salesPerformancePercent
        val differenceInQuantityTextView: TextView = binding.differenceInQuantitySold

        databaseHelper = DatabaseHelper(this.activity)

        val apprenticeRetriever = ApprenticeRetriever(date, requireContext())

        val salesDateButton = binding.salesDateButton
        val totalSalesMadeTextView: TextView = binding.salesMadeTextView

        val totalSalesMade = apprenticeRetriever.getTotalSalesMadeToday()
        totalSalesMadeTextView.text = currencyInstance.format(totalSalesMade)

        val salesQuantityTextView: TextView = binding.salesQuantityTextView
        salesQuantityTextView.text = apprenticeRetriever.getTotalQuantitySoldToday().toString()

        val salesNumberTextView: TextView = binding.salesNumberTextView
        salesNumberTextView.text = apprenticeRetriever.getTotalNumberOfSalesToday().toString()

        val customerNumberTextView: TextView = binding.customerNumberTextView
        customerNumberTextView.text = apprenticeRetriever.getNumberOfNewCustomersToday().toString()

        val dateSetListener = OnDateSetListener {
                _: DatePicker?, year: Int, month: Int, day: Int ->

                val calendar = Calendar.getInstance()
                calendar[Calendar.YEAR] = year
                calendar[Calendar.MONTH] = month
                calendar[Calendar.DAY_OF_MONTH] = day

                val pastDate = SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(calendar.time)
                val apprenticeRetrieverPast = ApprenticeRetriever(pastDate, requireContext())

                val dateOfPastSales = "Date of Sales: $pastDate"
                dateOfPastSalesTextView.text = dateOfPastSales

                val pastTotalSalesMade = apprenticeRetrieverPast.getTotalSalesMadeToday()
                pastSalesMadeTextView.text = currencyInstance.format(pastTotalSalesMade)

                pastSalesQuantityTextView.text = apprenticeRetrieverPast.getTotalQuantitySoldToday().toString()

                pastSalesNumberTextView.text = apprenticeRetrieverPast.getTotalNumberOfSalesToday().toString()

                pastCustomerNumberTextView.text = apprenticeRetrieverPast.getNumberOfNewCustomersToday().toString()

                val salesPerformanceValue = totalSalesMade - pastTotalSalesMade
                salesPerformanceValueTextView.text = currencyInstance.format(salesPerformanceValue)

                val decimalFormat = DecimalFormat("#.##")

                var salesPerformancePercent = 0.0
                if (totalSalesMade != 0.0) {
                    salesPerformancePercent =
                        decimalFormat.format((totalSalesMade - pastTotalSalesMade) / totalSalesMade * 100).toDouble()
                }

                if (java.lang.Double.isNaN(salesPerformancePercent)) {
                    salesPerformancePercent = 0.0
                }

                val salesPerformancePercentString = String.format(Locale.US, "%.0f%%", salesPerformancePercent)
                salesPerformancePercentTextView.text = salesPerformancePercentString

                val differenceInQuantity = apprenticeRetriever.getTotalQuantitySoldToday() - apprenticeRetrieverPast.getTotalQuantitySoldToday()
                differenceInQuantityTextView.text = differenceInQuantity.toString()
        }

        salesDateButton.setOnClickListener {
            val calendar = Calendar.getInstance()
            val year = calendar[Calendar.YEAR]
            val month = calendar[Calendar.MONTH]
            val day = calendar[Calendar.DAY_OF_MONTH]

            val datePickerDialog = DatePickerDialog(requireActivity(), dateSetListener, year, month, day)
            datePickerDialog.show()
        }

        return binding.root
    }

}